kernel/src/net/socket/unix/stream/connected.rs

```diff
@@ -57,69 +80,108 @@ impl Connected {
         let reader_events = self.reader.poll(mask, poller.as_deref_mut());
         let writer_events = self.writer.poll(mask, poller);
 
-        // FIXME: Check this logic later.
-        if reader_events.contains(IoEvents::HUP) || self.reader.is_shutdown() {
-            events |= IoEvents::RDHUP | IoEvents::IN;
-            if writer_events.contains(IoEvents::ERR) || self.writer.is_shutdown() {
-                events |= IoEvents::HUP | IoEvents::OUT;
-            }
-        }
-
-        events |= (reader_events & IoEvents::IN) | (writer_events & IoEvents::OUT);
-
-        events
+        combine_io_events(mask, reader_events, writer_events)
     }
 
     pub(super) fn register_observer(
         &self,
         observer: Weak<dyn Observer<IoEvents>>,
         mask: IoEvents,
     ) -> Result<()> {
-        if mask.contains(IoEvents::IN) {
-            self.reader.register_observer(observer.clone(), mask)?
-        }
-
-        if mask.contains(IoEvents::OUT) {
-            self.writer.register_observer(observer, mask)?
-        }
-
+        self.reader.register_observer(observer.clone(), mask)?;
+        self.writer.register_observer(observer, mask)?;
         Ok(())
     }
 
     pub(super) fn unregister_observer(
         &self,
         observer: &Weak<dyn Observer<IoEvents>>,
     ) -> Option<Weak<dyn Observer<IoEvents>>> {
-        let observer0 = self.reader.unregister_observer(observer);
-        let observer1 = self.writer.unregister_observer(observer);
-
-        if observer0.is_some() {
-            observer0
-        } else if observer1.is_some() {
-            observer1
-        } else {
-            None
+        let reader_observer = self.reader.unregister_observer(observer);
+        let writer_observer = self.writer.unregister_observer(observer);
+        reader_observer.or(writer_observer)
+    }
+}
+
+pub(super) fn combine_io_events(
+    mask: IoEvents,
+    reader_events: IoEvents,
+    writer_events: IoEvents,
+) -> IoEvents {
+    let mut events = IoEvents::empty();
+
+    if reader_events.contains(IoEvents::HUP) {
+        // The socket is shut down in one direction: the remote socket has shut down for
+        // writing or the local socket has shut down for reading.
+        events |= IoEvents::RDHUP | IoEvents::IN;
+
+        if writer_events.contains(IoEvents::ERR) {
+            // The socket is shut down in both directions. Neither reading nor writing is
+            // possible.
+            events |= IoEvents::HUP;
+        }
+    }
+
+    events |= (reader_events & IoEvents::IN) | (writer_events & IoEvents::OUT);
+
+    events & (mask | IoEvents::ALWAYS_POLL)
+}
+
+struct AddrCommon {
+    first: SpinLock<Option<UnixSocketAddrBound>>,
+    second: SpinLock<Option<UnixSocketAddrBound>>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum AddrEndpoint {
+    First,
+    Second,
+}
+
+struct AddrView(Arc<AddrCommon>, AddrEndpoint);
+
```

Comment:
Actually, there is a tradeoff between code readability and less heap allocation/more cache friendliness.

Since you found the original implementation hard to understand, your proposal is also acceptable to me.